contracts/fatex/FateReward/MembershipWithReward.sol

```diff
@@ -0,0 +1,268 @@
+                depositedUsers[_pid],
+                _user
+            );
+            if (userIndex == depositedUsers[_pid].length) {
+                // not recored yet (first deposit)
+                depositedUsers[_pid].push(_user);
+                userMembershipInfo[_pid][_user] = MembershipInfo({
+                    firstDepositBlock: currentBlockNumber,
+                    rankedNumber: 2 ** 256 - 1,
+                    lastWithdrawBlock: currentBlockNumber
+                });
+            }
+        }
+    }
+
+    /// @dev calculate Points earned by this user
+    function userPoints(uint256 _pid, address _user) public view returns (uint256 points){
+        points = _getBlocksOfPeriod(
+            _pid,
+            _user,
+            true
+        ) * POINTS_PER_BLOCK;
+    }
+
+    /// @dev rank with Points and set rankedNumber to each user
+    function rank(uint256 _pid) public returns (uint256) {
+        if (isFatePool[_pid]) {
+            address[] memory pool_deposited_user_list = depositedUsers[_pid];
+            uint256 deposited_user_counts = pool_deposited_user_list.length;
+            require(deposited_user_counts > 0, "rank: no_deposited_users_yet");
+            
+
+            uint256[] memory pointsList = new uint256[](deposited_user_counts);
+            for(uint i = 0; i < deposited_user_counts; i++) {
+                pointsList[i] = userPoints(_pid, pool_deposited_user_list[i]);
+            }
+
+            uint256[] memory sortedPointsList = RankedArray.sort(pointsList);
+            for (uint i = 0; i < deposited_user_counts; i++) {
+                address userAddr = depositedUsers[_pid][i];
+                MembershipInfo memory membership = userMembershipInfo[_pid][userAddr];
+
+                userMembershipInfo[_pid][userAddr] = MembershipInfo({
+                    firstDepositBlock: membership.firstDepositBlock,
+                    rankedNumber: RankedArray.getIndex(sortedPointsList, pointsList[i]),
+                    lastWithdrawBlock: membership.lastWithdrawBlock
+                });
+            }
+
+        }
+    }
+
+    function _getBlocksOfPeriod(
+        uint256 _pid,
+        address _user,
+        bool _isDepositPeriod
+    ) internal view returns (uint256 blocksOfPeriod) {
+        if (isFatePool[_pid]) {
+            uint256 currentBlockNumber = block.number;
+            uint256 endBlock = currentBlockNumber > epochEndBlock ? epochEndBlock : currentBlockNumber;
+            MembershipInfo memory membership = userMembershipInfo[_pid][_user];
+            uint256 startBlock = _isDepositPeriod ? membership.firstDepositBlock : membership.lastWithdrawBlock;
+            
+            if (endBlock >= startBlock) {
+                blocksOfPeriod = endBlock - startBlock;
+            }
+        }
+    }
+
+    /// @dev calculate lockedRewardsFees as percent that will be sent to the rewardController
+    function _getLockedRewardsFeePercent(
```

Comment:
Add better documentation - is the `percent` returned the percent kept by the user or the percent paid as fees from their rewards?